quickcheck-dynamic/src/Test/QuickCheck/StateModel.hs

```diff
@@ -327,31 +383,47 @@ instance Show state => Show (Annotated state) where
 initialAnnotatedState :: StateModel state => Annotated state
 initialAnnotatedState = Metadata mempty initialState
 
-computePrecondition :: StateModel state => Annotated state -> Action state a -> Bool
-computePrecondition s a =
-  all (\(Some v) -> v `isWellTyped` vars s) (getAllVariables a)
-    && precondition (underlyingState s) a
+actionWithPolarity :: (StateModel state, Eq (Action state a)) => Annotated state -> Action state a -> ActionWithPolarity state a
+actionWithPolarity s a =
+  let p
+        | precondition (underlyingState s) a = PosPolarity
+        | negativePrecondition (underlyingState s) a = NegPolarity
+        | otherwise = PosPolarity
```

Comment:
Does this not mean we just discard the result of `precondition`? If it fails, and the negativePrecondition fails, the action is still considered `Positive` 🤔

Reply:
Ah no, because we don't use that to actually filter the action.

Reply:
This semantics is really important to make it easy to add negative testing. See the Regsitry example where the negative precondition is always True - its a one line change that turns negative testing on. That example also adds a negative postcondition to check the error result- but thats not strictly necessary.

This was my main design concern. You should be able to qc-d without knowing anything about negative testing and turning it on should be as simple as possible once you have a working model.